### PR TITLE
Fix incorrect math surrounding text transforms in CGContextDrawGlyphRun

### LINKED ISSUE #1113: WOC Catalog CTLineCreateTruncatedLine test incorrect
Lines are being drawn ordered upside-down

### LINKED ISSUE #1135: Fix conflation of text and context transformations in CoreText/Graphics

### PR DESCRIPTION
**Pull request by ms-jihua**
The text transforms were being incorrectly applied, directly concatenated onto the final transform.
The text transform must be applied in text space (centered on the glyph run, inverted)

Fixes #1113
Fixes #1135

### CODE REVIEW

File: Frameworks/CoreGraphics/CGContextCairo.mm [hunk 1]
```diff
@@ -2011,7 +2011,32 @@
...
+    // Flip and translate by the difference between the center and the baseline, apply text transforms, then flip and translate back
+    DWRITE_FONT_METRICS fontMetrics;
+    glyphRun->fontFace->GetMetrics(&fontMetrics);
+    float baselineCenterDelta = (fontMetrics.ascent - fontMetrics.descent) * glyphRun->fontEmSize / fontMetrics.designUnitsPerEm / 2;
+
+    // Transform to text space
+    // Technically there should be a horizontal translation to the center as well,
+    // but it's to the center of _each individual glyph_, as the reference platform applies the text matrix to each glyph individually
+    // Uncertain whether it's ever going to be worth it to support this using DWrite, so just ignore it for now
+    transform = CGAffineTransformConcat(CGAffineTransformMake(1, 0, 0, -1, 0, baselineCenterDelta), transform);
```
Comment (rajsesh-msft) [inline Frameworks/CoreGraphics/CGContextCairo.mm]: You could skip the initial CGAffineTransformConcat since transform is identity.

Comment (rajsesh-msft) [inline Frameworks/CoreGraphics/CGContextCairo.mm]: Or rather: CGAffineTransform transform = CGAffineTransformMake(...);

---
In reply to: [84172077](https://github.com/Microsoft/WinObjC/pull/1188#discussion_r84172077) [](ancestors = 84172077)

**Comment by DHowett-MSFT** [comment]
So this anchors the bottom left of the text at the text position?
If the context is transformed `1 0 0 -1 0 0`, it then anchors the top left of the upside-down text at the text position?

**Comment by DHowett-MSFT** [comment]
:shipit:

**Comment by ms-jihua** [comment]
More or less yes. At the baseline instead of the bottom/top left corner though.

---
In reply to: [254951164](https://github.com/Microsoft/WinObjC/pull/1188#issuecomment-254951164) [](ancestors = 254951164)

**Pull request merged**